File: scripts/test_t661_fill.js
// scripts/test_t661_fill.js
// Node script to test T661 field positions by overlaying colored text on the actual PDF

import { readFile, writeFile } from "fs/promises";
import { PDFDocument, rgb, StandardFonts } from "pdf-lib";

const mappingPath = "/Users/jj/rambo2/supabase/field_mappings/t661_mapping.json";
const templatePath = "/Users/jj/rambo2/supabase/templates/t661-20e.pdf";
const outputPath = "/Users/jj/rambo2/test_output_t661.pdf";

// Sample test data
const testData = {
    company_name: "ACME RESEARCH CORP",
    business_number: "123456789RC0001",
    tax_year: "2024-12-31",
    project_id: "P001",
    project_title: "AI-Powered Quality Control System",
    project_summary: "This project develops a novel computer vision system using deep learning for real-time defect detection in manufacturing.",
    scientific_advancement: "We pursue technological advancement in automated defect detection using novel neural network architectures that can achieve higher accuracy with less training data.",
    base_knowledge: "Existing computer vision systems lack accuracy for micro-defects below 0.1mm. Current ML models require 10,000+ labeled samples per defect type.",
    uncertainties: "Whether our proposed hybrid CNN-Transformer architecture can achieve 99% accuracy with only 500 training samples per defect type remains scientifically uncertain.",
    work_description: "We conducted systematic experiments: 1) Novel architecture design combining CNNs with attention mechanisms, 2) Advanced data augmentation techniques, 3) Transfer learning from pre-trained models."
};

(async () => {
    try {
        // Load mapping and PDF
        const mapping = JSON.parse(await readFile(mappingPath, "utf-8"));
        const pdfBytes = await readFile(templatePath);
        const pdfDoc = await PDFDocument.load(pdfBytes);
        const font = await pdfDoc.embedFont(StandardFonts.Helvetica);
        const fontBold = await pdfDoc.embedFont(StandardFonts.HelveticaBold);

        console.log(`PDF has ${pdfDoc.getPageCount()} pages`);
        console.log("Adding test text overlays (in RED for visibility)...\n");

        // Overlay text on pages
        for (const [sectionName, fields] of Object.entries(mapping.fields)) {
            for (const [fieldName, fieldConfig] of Object.entries(fields)) {
                const llmKey = fieldConfig.llm_source || fieldName;
                const value = testData[llmKey];

                if (!value) continue;

                const page = pdfDoc.getPage(fieldConfig.page);
                const textValue = String(value);

                if (fieldConfig.multiline && fieldConfig.maxWidth) {
                    // Handle multi-line text
                    const words = textValue.split(" ");
                    const lines = [];
                    let currentLine = "";

                    for (const word of words) {
                        const testLine = currentLine ? `${currentLine} ${word}` : word;
                        const width = font.widthOfTextAtSize(testLine, fieldConfig.fontSize);

                        if (width > fieldConfig.maxWidth && currentLine) {
                            lines.push(currentLine);
                            currentLine = word;
                        } else {
                            currentLine = testLine;
                        }
                    }
                    if (currentLine) lines.push(currentLine);

                    const maxLines = fieldConfig.maxHeight
                        ? Math.floor(fieldConfig.maxHeight / (fieldConfig.fontSize * 1.2))
                        : lines.length;

                    for (let i = 0; i < Math.min(lines.length, maxLines); i++) {
                        page.drawText(lines[i], {
                            x: fieldConfig.x,
                            y: fieldConfig.y - i * fieldConfig.fontSize * 1.2,
                            size: fieldConfig.fontSize,
                            font: font,
                            color: rgb(1, 0, 0), // RED for testing visibility
                        });
                    }

                    console.log(
                        `✓ Page ${fieldConfig.page + 1}: ${fieldName} (multiline, ${lines.length} lines) at (${fieldConfig.x}, ${fieldConfig.y})`
                    );
                } else {
                    // Single-line text
                    const displayText = textValue.substring(0, 60);
                    page.drawText(displayText, {
                        x: fieldConfig.x,
                        y: fieldConfig.y,
                        size: fieldConfig.fontSize,
                        font: font,
                        color: rgb(1, 0, 0), // RED for testing visibility
                    });

                    console.log(
                        `✓ Page ${fieldConfig.page + 1}: ${fieldName} at (${fieldConfig.x}, ${fieldConfig.y})`
                    );
                }
            }
        }

        // Save
        const filledBytes = await pdfDoc.save();
        await writeFile(outputPath, filledBytes);

        console.log(`\n✅ Test PDF created: ${outputPath}`);
        console.log("\n📋 Next steps:");
        console.log("1. Open test_output_t661.pdf");
        console.log("2. Compare with your screenshots");
        console.log("3. For each RED text field, tell me:");
        console.log("   - 'field_name: move up/down X points'");
        console.log("   - 'field_name: move left/right X points'");
        console.log("   - 'field_name: looks good'");
    } catch (e) {
        console.error("Error:", e);
        process.exit(1);
    }
})();

File: scripts/test_t661_fill.py
#!/usr/bin/env python3
"""
Simple test script to fill T661 with sample data and generate a test PDF.
This allows visual verification of field positions without deploying to Supabase.
"""

import json
from pathlib import Path

try:
    from PyPDF2 import PdfReader, PdfWriter
    from reportlab.pdfgen import canvas
    from reportlab.lib.pagesizes import letter
    from reportlab.pdfbase import pdfmetrics
    from reportlab.pdfbase.ttfonts import TTFont
    import io
except ImportError:
    print("Error: Required packages not installed")
    print("Run: pip3 install PyPDF2 reportlab")
    exit(1)

# Load field mapping
mapping_path = Path("/Users/jj/rambo2/supabase/field_mappings/t661_mapping.json")
with open(mapping_path) as f:
    mapping = json.load(f)

# Sample test data
test_data = {
    "company_name": "Acme Research Corp",
    "business_number": "123456789RC0001",
    "tax_year": "2024-12-31",
    "project_id": "P001",
    "project_title": "Development of Advanced AI-Powered Quality Control System",
    "project_summary": "This project aims to develop a novel computer vision system that uses deep learning to detect manufacturing defects in real-time.",
    "scientific_advancement": "We are pursuing technological advancement in automated defect detection using novel neural network architectures.",
    "base_knowledge": "Existing computer vision systems lack accuracy for micro-defects. Current ML models require extensive labeled data.",
    "uncertainties": "Whether our proposed architecture can achieve 99% accuracy with limited training data remains unclear.",
    "work_description": "We conducted systematic experiments including: 1) Architecture design, 2) Data augmentation techniques, 3) Transfer learning approaches."
}


# Load template PDF
template_path = Path("/Users/jj/rambo2/supabase/templates/t661-20e.pdf")

# Create output PDF with text overlays
output_path = Path("/Users/jj/rambo2/test_output_t661.pdf")
can = canvas.Canvas(str(output_path), pagesize=letter)

# We'll just draw on page 1 and 2 for initial testing
reader = PdfReader(template_path)
total_pages = len(reader.pages)

print(f"Template PDF has {total_pages} pages")
print("Drawing test data on pages 1-2...")

# For each page, draw the template as background then add text
for page_num in range(min(2, total_pages)):  # Just test first 2 pages
    # Start a new page
    if page_num > 0:
        can.showPage()
    
    # Draw fields for this page
    for section_name, fields in mapping["fields"].items():
        for field_name, field_config in fields.items():
            if field_config["page"] == page_num:
                llm_key = field_config.get("llm_source", field_name)
                value = test_data.get(llm_key, "")
                
                if value:
                    can.setFont("Helvetica", field_config["fontSize"])
                    # Truncate long text for testing
                    text = str(value)[:80] if not field_config.get("multiline") else str(value)[:200]
                    can.drawString(field_config["x"], field_config["y"], text)
                    print(f"  Page {page_num + 1}: {field_name} at ({field_config['x']}, {field_config['y']})")

can.save()


print(f"✅ Test PDF created: {output_path}")
print("Open this PDF and compare with your screenshots to see if fields are in the right positions.")
print("\nThen tell me adjustments like:")
print("  - 'company_name is too high, move down 20 points'")
print("  - 'project_title is too far left, move right 30 points'")
